### DISCUSSION TITLE
Function support for __webpack_public_path__

### DISCUSSION THREAD

**Discussion by Iswaryasl** [webpack/webpack#17327] [Q&A]
## Feature request

I am requesting the addition of function support for the __webpack_public_path__ variable in webpack. Currently, __webpack_public_path__ is a variable that allows for setting a static public path at compile time. However, there are scenarios where the ability to dynamically compute the public path at runtime would be highly beneficial.

**What is the expected behavior?**

This would allow developers to provide a function that dynamically computes and returns the appropriate public path at runtime based on custom logic or environment variables.

For example:
```
__webpack_public_path__ = (fileName) => {
  // Custom logic to determine the public path based on the environment or configuration
  if(fileName.indexOf('javascript') != -1) {
    return "/assets/js"
  } else {
    return process.env.BASE_URL || '/';
  }
};
```

By introducing function support, webpack would call the function whenever it needs to determine the public path, allowing for runtime calculations and dynamic behavior.

**What is motivation or use case for adding/changing the behavior?**

The current limitation of __webpack_public_path__ being a static variable poses challenges in certain use cases. For example, when deploying an application to different environments or serving assets from different CDNs, it would be advantageous to calculate the public path dynamically based on the current environment or configuration. This would enable more flexibility and ease of deployment for different environments without the need for manual configuration changes.

**How should this be implemented in your opinion?**

If function support is implemented for __webpack_public_path__, instead of directly assigning the public path, we could call a public path assignment function with the fileName or chunkId as an argument (__webpack_require__.p(chunkId)). This would allow us to dynamically compute and retrieve the appropriate public path based on the specific chunkId or fileName.
For example
```
var fullhref = __webpack_require__.p(chunkId) + href;
```
**Are you willing to work on this yourself?**
No

**Comment by alexander-akait** [comment]
You can do it in runtime:
```js
function setPublicPath(path) {
  if(path.indexOf('javascript') != -1) {
    __webpack_public_path__ = "/assets/js"
  } else {
    __webpack_public_path__ = path || '/';
  }
}
```

Sorry, we don't want to implement this because automatic public path is recommended, if you have exotic usage you can implement logic above and change it when you need and reset it, also there are some limitation about this, because you can define multiple functions in different places, our runtime code written for static values and some other things.

Also you can setup publicPath on the entripoints level, you can use function to get different URLs for different assets

I can accept your PRs, but if you

> Are you willing to work on this yourself?
> No

It will most likely never be implemented.

**Reply by Iswaryasl** [reply]
> You can do it in runtime:
> 
> ```js
> function setPublicPath(path) {
>   if(path.indexOf('javascript') != -1) {
>     __webpack_public_path__ = "/assets/js"
>   } else {
>     __webpack_public_path__ = path || '/';
>   }
> }
> ```
> 
> Sorry, we don't want to implement this because automatic public path is recommended, if you have exotic usage you can implement logic above and change it when you need and reset it, also there are some limitation about this, because you can define multiple functions in different places, our runtime code written for static values and some other things.
> 
> Also you can setup publicPath on the entripoints level, you can use function to get different URLs for different assets
> 
> I can accept your PRs, but if you
> 
> > Are you willing to work on this yourself?
> > No
> 
> It will most likely never be implemented.

I m ready to work on tis

**Comment by Iswaryasl** [comment]
Can you please elaborate how to use it in runtime.

On Mon, May 29, 2023 at 1:56 AM Alexander Akait ***@***.***>
wrote:

> You can do it in runtime:
>
> function setPublicPath(path) {
>   if(path.indexOf('javascript') != -1) {
>     __webpack_public_path__ = "/assets/js"
>   } else {
>     __webpack_public_path__ = path || '/';
>   }}
>
> Sorry, we don't want to implement this because automatic public path is
> recommended, if you have exotic usage you can implement logic above and
> change it when you need and reset it, also there are some limitation about
> this, because you can define multiple functions in different places, our
> runtime code written for static values and some other things.
>
> Also you can setup publicPath on the entripoints level, you can use
> function to get different URLs for different assets
>
> I can accept your PRs, but if you
>
> Are you willing to work on this yourself?
> No
>
> It will most likely never be implemented.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/webpack/webpack/issues/17274#issuecomment-1566252271>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGEUKUYLUCBBTCGXK2XGTHDXIOYF5ANCNFSM6AAAAAAYRTNVIQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Regards,

           Iswarya S L
           Member Technical Staff
           Zoho Corporation,Chennai.

**Comment by alexander-akait** [comment]
@Iswaryasl 

> Can you please elaborate how to use it in runtime.

Use what? You just define own function and run `setPublicPath` when you need it.

Also note - `__webpack_public_path__` will not work for static assets, like CSS, HTML and etc, so dynamic overriding `__webpack_public_path__` should be the last solution if nothing was found. So I recommend reconsider your configuration

**Comment by Iswaryasl** [comment]
I tried to use setPublicPath like below in my webpack.config.js
<img width="789" alt="Screenshot 2023-06-06 at 11 21 28 AM" src="https://github.com/webpack/webpack/assets/25773395/f1ea4896-f947-4602-9bc4-6a6fd17d3264">

I m getting the error unknown configuration

> Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema configuration has an unknown property 'setPublicPath'. 

Can you please help on this

**Reply by alexander-akait** [reply]
You should to use this in source code before places where you can change public path, note - it will work only for JS, not for HTML or CSS assets, because there is no ability to use variables/functions in HTML and CSS
